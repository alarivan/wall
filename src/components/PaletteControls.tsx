import React from 'react';
import styled from 'styled-components';
import Palette from './Palette';
import { TColor } from '../types';

export interface Props {
  colors: TColor[];
  current: TColor;
  colorPicker: boolean;
  onColorClick: (color: TColor) => void;
  onClearClick: () => void;
  onColorPickerClick: () => void;
}

const StyledPaletteControls = styled.div`
  padding: 0 0.5rem;
  display: grid;
  grid-template-columns: 1fr;
  grid-gap: 2px;
  align-content: start;
`;

const PaletteControls: React.FC<Props> = ({
  current,
  colors,
  colorPicker,
  onColorClick,
  onClearClick,
  onColorPickerClick,
}) => {
  function handleEraserClick() {
    onColorClick(undefined);
  }

  return (
    <StyledPaletteControls>
      <Palette onColorClick={onColorClick} current={current} colors={colors} />
      <button data-testid='palette-controls-eraser' onClick={handleEraserClick}>
        eraser
      </button>
      <button data-testid='palette-controls-clear' onClick={onClearClick}>
        clear
      </button>
      <button
        data-testid='palette-controls-color-picker'
        disabled={colorPicker}
        onClick={onColorPickerClick}
      >
        pick color
      </button>
    </StyledPaletteControls>
  );
};

export default PaletteControls;
