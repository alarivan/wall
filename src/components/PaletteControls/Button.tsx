import styled from 'styled-components';
import { darken, readableColor } from 'polished';

function getTextColor(defaultColor: string, color?: string): string {
  return color ? readableColor(color) : readableColor(defaultColor);
}

interface Props {
  bgColor?: string;
}

const Button = styled.button<Props>`
  cursor: pointer;
  padding: 0.25rem 0.5rem;
  margin-bottom: 0.25rem;
  border: none;
  font-weight: 500;
  text-transform: uppercase;
  box-shadow: 1px 1px 5px 0px rgba(0, 0, 0, 0.75);
  transition: background-color 200ms ease-in;

  color: ${({ theme, bgColor }) => getTextColor(theme.secondary, bgColor)};
  background-color: ${({ theme, bgColor }) =>
    bgColor ? bgColor : theme.secondary};
  :hover {
    background-color: ${({ theme }) => darken(0.1, theme.secondary)};
  }
`;

export default Button;
