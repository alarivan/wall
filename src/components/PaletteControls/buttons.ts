import withTippy, { TippyPropsLite } from '../common/withTippy';
import Button from './Button';

const options: TippyPropsLite = {
  placement: 'right',
};

export const EraserButton = withTippy(Button, 'Eraser', options);

export const BackgroundButton = withTippy(
  Button,
  'Change background color',
  options,
);

export const ColorPickerButton = withTippy(
  Button,
  'Pick color from grid',
  options,
);

export const CustomColorButton = withTippy(
  Button,
  'Choose custom color',
  options,
);
