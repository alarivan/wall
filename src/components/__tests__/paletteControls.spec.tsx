import React from 'react';
import { render, cleanup, fireEvent } from '@testing-library/react';
import PaletteControls, { Props } from '../PaletteControls';

afterEach(cleanup);

const props: Props = {
  colors: ['red'],
  current: 'red',
  colorPicker: false,
  onColorClick: jest.fn(),
  onClearClick: jest.fn(),
  onColorPickerClick: jest.fn(),
};

describe('PaletteControls', () => {
  it('renders correctly', () => {
    const { asFragment, getByTestId } = render(<PaletteControls {...props} />);

    expect(getByTestId('palette-controls-color-picker')).not.toHaveAttribute(
      'disabled',
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('disables color picker when it is active', () => {
    const { getByTestId } = render(
      <PaletteControls {...props} colorPicker={true} />,
    );

    expect(getByTestId('palette-controls-color-picker')).toHaveAttribute(
      'disabled',
    );
  });

  it('calls onClick on correct buttons', () => {
    const { getByTestId } = render(<PaletteControls {...props} />);

    fireEvent.click(getByTestId('palette-controls-eraser'));
    fireEvent.click(getByTestId('palette-controls-clear'));
    fireEvent.click(getByTestId('palette-controls-color-picker'));

    expect(props.onColorClick).toHaveBeenCalledTimes(1);
    expect(props.onClearClick).toHaveBeenCalledTimes(1);
    expect(props.onColorPickerClick).toHaveBeenCalledTimes(1);
  });
});
