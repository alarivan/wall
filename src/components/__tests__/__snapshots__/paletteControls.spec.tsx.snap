// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PaletteControls renders correctly 1`] = `
<DocumentFragment>
  .c2 {
  background: red;
  border-style: solid;
  border-color: red;
  border-width: 0 1px 1px 0;
  -webkit-transition: -webkit-transform 100ms ease-in-out;
  -webkit-transition: transform 100ms ease-in-out;
  transition: transform 100ms ease-in-out;
  cursor: pointer;
  box-shadow: 1px 1px 5px 0px rgba(0,0,0,0.75);
  border-style: solid;
  border-color: black;
  border-width: 0;
  border-radius: 50%;
  -webkit-transition: border-radius 50ms ease-in-out;
  transition: border-radius 50ms ease-in-out;
}

.c2:hover {
  border-width: 1px;
  -webkit-transform: scale(1.1);
  -ms-transform: scale(1.1);
  transform: scale(1.1);
}

.c2:hover {
  border-width: 0;
  border-radius: 50%;
}

.c1 {
  display: grid;
  grid-gap: 2px;
  grid-template-columns: repeat(3,20px);
  grid-auto-rows: 20px;
  -webkit-align-self: start;
  -ms-flex-item-align: start;
  align-self: start;
  background: #c2c2d1;
  padding: 0.25rem;
}

.c0 {
  padding: 0 0.5rem;
  display: grid;
  grid-template-columns: 1fr;
  grid-gap: 2px;
  -webkit-align-content: start;
  -ms-flex-line-pack: start;
  align-content: start;
}

<div
    class="c0"
  >
    <div
      class="c1"
    >
      <div
        class="c2"
        data-testid="palette-cell"
      />
    </div>
    <button
      data-testid="palette-controls-eraser"
    >
      eraser
    </button>
    <button
      data-testid="palette-controls-clear"
    >
      clear
    </button>
    <button
      data-testid="palette-controls-color-picker"
    >
      pick color
    </button>
  </div>
</DocumentFragment>
`;
